Fix incorrect light values on blend import

neonmoe opened the pull request:
This commit changes the Blender glTF export process to output light intensities in "unitless" lighting mode, which brings the brightness levels down to usable ranges when not using physical lighting units.

Physical lighting units could be fixed by e.g. using the default lighting mode, and passing the exported lumens into Godot's own light intensity parameter (which is also in lumens). That's not included in this PR due to it being a less clear fix, and Blender possibly changing their watt-to-lumen conversion in the near future.

Additionally, this sets the light attenuation for glTF (and thus Blender) imports to 2.0 by default, to match the attenuation seen in Cycles and EEVEE.

Fixes #92168.

### Screenshots

The screenshots using physical light units use camera attributes tuned so that the default Blender cube looks right (3 f-stop, 100 Hz shutter speed, 1200 ISO sensitivity). Alas, the results are still somewhat inconsistent between light types when using physical light units, leading to especially wrong results for directional lights.

The non-physical light units screenshots are using the default camera attributes' exposure, which seems to lead to a lightly brighter look than Blender by default, but it's not too far off, and the light intensities between light types are consistent.

| Blender Cycles | Blender EEVEE | Godot (physical light units) | Godot (non-physical light units) |
|---|---|---|---|
| ![render-blender-4 4 3-cycles](https://github.com/user-attachments/assets/6c789178-a331-427f-b298-ba1d1574b965) | ![render-blender-4 4 3-eevee](https://github.com/user-attachments/assets/02fc9e63-317d-49bf-bace-877eb7abcf1e) | ![render-godot-PR-forward+](https://github.com/user-attachments/assets/221f0236-ac0b-4991-aea6-70c58379bef5) | ![render-godot-PR-forward+-non-physical](https://github.com/user-attachments/assets/b37e90e4-3566-4259-be30-873989d2efd2) |
| ![render-blender-4 4 3-cycles](https://github.com/user-attachments/assets/cff5022c-74eb-430d-8728-22446da76169) | ![render-blender-4 4 3-eevee](https://github.com/user-attachments/assets/4e53abaf-cb08-4637-903c-0d564333109f) | ![render-godot-PR-forward+](https://github.com/user-attachments/assets/7e2414c9-330a-4440-af4f-13e37e493f1b) | ![render-godot-PR-forward+-non-physical](https://github.com/user-attachments/assets/708faafb-751c-44d3-a570-6f0f25280233) |
| ![render-blender-4 4 3-cycles](https://github.com/user-attachments/assets/ec3bcd32-9bc4-48d0-bd18-dce77ccbbd63) | ![render-blender-4 4 3-eevee](https://github.com/user-attachments/assets/a3408ac8-3c2f-4272-8d9d-5caf68dada25) | ![render-godot-PR-forward+](https://github.com/user-attachments/assets/d13dc790-4f59-4080-8d78-b4f9e4bb588b) | ![render-godot-PR-forward+-non-physical](https://github.com/user-attachments/assets/6a488a51-4999-4a27-b54b-b31d44307348) |
| ![render-blender-4 4 3-cycles](https://github.com/user-attachments/assets/cde14727-4ecd-4316-9d5b-33c1de83e428) | ![render-blender-4 4 3-eevee](https://github.com/user-attachments/assets/aac30a2d-91e1-43f7-99e3-dd6e89445922) | ![render-godot-PR-forward+](https://github.com/user-attachments/assets/04fea6ba-99d6-40ad-adde-8d5ef2982b38) | ![render-godot-PR-forward+-non-physical](https://github.com/user-attachments/assets/b43687fe-b898-47af-b6e0-bf24085ecbd0) |

(The Godot screenshots reflect commit 54fd2853b17c5633b1553389d9aeac37f72845da.)

The first scene is just the default Blender cube with the background taken out to ease calibration. The second one has all the three light types, to validate that they look appropriate relative to each other. The third one has a single directional light, which is The Sun levels of bright, except it has an almost black color to avoid overexposing the default ("indoors") exposure. The fourth one is the Blender scene included in the issue's MRP, with the background set to black (since that can't be accounted for by this import process, and the default Godot background is too bright for this exposure with physical light units).

The point of the three scenes is to move around the lights, use different intensities, and then see if the same exposure settings lead to matching results with Blender. Evidently the physical lighting path ends up having problematic directional lights, but everything else certainly looks better than without this PR (where all these are just pure white when facing towards a light).

Here's the project including all the scenes that those screenshots are from: 
[light-tests.zip](https://github.com/user-attachments/files/21800538/light-tests.zip)

neonmoe committed:
Fix light values on blend import when using non-physical light units

neonmoe commented:
Ah, nevermind. Changed into a draft, noticed that the MRP in the issue isn't actually fixed by this, needs further investigation :smile:

neonmoe commented on modules/gltf/editor/editor_scene_importer_blend.cpp:
```diff
@@ -325,12 +330,47 @@ Node *EditorSceneFormatImporterBlend::import_scene(const String &p_path, uint32_
+	TypedArray<Light3D> light_nodes = scene->find_children("*", "Light3D");
+	for (Variant light_node : light_nodes) {
+		Light3D *light = cast_to<Light3D>(light_node);
+		// Blender has physically based attenuation, so this is needed to get comparable lighting results.
+		light->set_param(Light3D::PARAM_ATTENUATION, 2.0);
```
Ah, yeah. That's probably better to do in another PR then, rather keep this one focused on the very obvious issue of the light energy/intensity values.

passivestar replied:
Not sure if it is actually, the discussion in that proposal was about changing the default value in nodes which would immediately break existing projects. But this PR changes import logic, not the default. Strictly speaking it's already breaking import compatibility by changing light intensity, so maybe it makes sense to fix the attenuation as well

fire replied:
I agree with @passivestar

neonmoe commented on modules/gltf/editor/editor_scene_importer_blend.cpp:
```diff
@@ -325,12 +330,47 @@ Node *EditorSceneFormatImporterBlend::import_scene(const String &p_path, uint32_
+				light->set_param(Light3D::PARAM_ENERGY, blender_light_units / 10.0);
+			} else {
+				light->set_param(Light3D::PARAM_ENERGY, blender_light_units / 3.0);
+			}
```
Turns out they ended up a bit brighter than Blender when comparing default exposure to default exposure, but I think that's very acceptable given how simple of a fix this was.

neonmoe commented:
Rebased after making changes based on reviews, and updated the screenshots to match.

fire reviewed (commented):
I will approve after I get a second opinion from @allenwp

allenwp commented:
> I will approve after I get a second opinion from @allenwp 

Thanks for the mention, I always appreciate it :)

Should we wait until KhronosGroup/glTF-Blender-IO#2554 is merged before locking in the luminous efficacy value used (177.83)? Also, in the comment where this constant is described, it might be good to link this source PR or related source code in other repositories.

I'd like to hear thoughts from @will-ca and @RevoluPowered who have already discussed some of this in #73624

I would love to dive deep into this, but I don't have the time in the next while to familiarize myself with Blender and its standards.

neonmoe commented:
After writing out [this comment on the glTF-specific issue](https://github.com/godotengine/godot/issues/73624#issuecomment-3046288978), I think this might actually be best fixed for physical light units by letting Blender export using the "spec" lighting mode, passing the glTF light intensity values to our intensity parameter (and multiplying spot and omnilights by 4*pi), and leaving the energy parameter as 1. Then Godot would be working according to the glTF spec, and any luminous efficacy related decisions would be up to Blender. Currently it'd export lights using 683lm/W, which would be a little bright compared to 177.82lm/W, but still definitely in a usable range, compared to passing the 683-scale values into the energy parameter (which gets multiplied by 1000lm or 100000lx due to the default intensity), as Godot is currently doing.

That'd also fix #73624.

And once Blender switches to 177.83lm/W, Godot wouldn't need any changes to reflect that.

fire commented:
Any idea when Blender switches to 177.83lm/W?

aaronfranke commented on modules/gltf/extensions/gltf_light.cpp:
```diff
@@ -182,6 +182,7 @@ Light3D *GLTFLight::to_node() const {
 		return nullptr;
 	}
 	light->set_color(color.linear_to_srgb());
+	light->set_param(Light3D::PARAM_ATTENUATION, 2.0);
```
Should we do anything with this value on export? We want to ensure that if a glTF file is exported from Godot, it will be imported again with at least roughly the same appearance.

aaronfranke reviewed (commented):
> Would it make sense to drop the use_physical_light_units half of this PR, and just merge the unitless fix?

It would be unfortunate if Godot depended on one behavior from Blender which changed out from under us in a later Blender update, so I would prefer that issue to be resolved first on Blender's side before we add that feature. So yes, I'd drop it from this PR, and we can add it in later. Or, you can open another PR with a more minimal fix, and we can leave this one for later.

neonmoe commented:
> So yes, I'd drop it from this PR, and we can add it in later. Or, you can open another PR with a more minimal fix, and we can leave this one for later.

I'll just do the minimal fix here, better to have a fresh go at the proper fix later. I've now cleaned up the PR to just have the unitless fix, and updated the screenshots. The results are of course worse for the physical light units path, but at least the non-physical light units screenshots look almost correct, with a very small change to the code.
